lib/spack/docs/advanced_topics.rst

```diff
@@ -0,0 +1,223 @@
+
+Spack lets you specify compilers on the CLI with, e.g., ``%gcc`` or ``%c,cxx=clang %fortran=gcc``, and you can specify flags with ``cflags``, ``cxxflags``, and ``fflags``.
+Depending on how complex your compiler setup is, it can be cumbersome to specify all of your preferences on the CLI.
+Spack has a special type configuration called ``toolchains``, which let you encapsulate the configuration for compilers, other libraries, and flags into a single name that you can reference as though it were one option.
+
+Toolchains are referenced by name like a direct dependency, using the ``%`` sigil.
+They are defined under the ``toolchains`` section of the configuration:
+
+.. code-block:: yaml
+
+   toolchains:
+     llvm_gfortran:
+     - spec: cflags=-O3
+     - spec: '%[virtuals=c] llvm'
+       when: '%c'
+     - spec: '%[virtuals=cxx] llvm'
+       when: '%cxx'
+     - spec: '%[virtuals=fortran] gcc'
+       when: '%fortran'
+
+When used, the toolchain above enforces using ``llvm`` for the C and C++ languages, and ``gcc`` for Fortran, when these languages are needed.
+It also adds ``cflags=-O3`` unconditionally.
+
+Using *conditional dependencies*, when constraining a virtual provider in a toolchain, is the key to make the toolchain itself widely applicable.
+For instance, a toolchain written that way can be used to install a list of specs using :ref:`environment-spec-matrices`:
+
+.. code-block:: yaml
+
+   specs:
+   - matrix:
+     - [kokkos, hdf5, py-scipy]
+     - ["%llvm_gfortran"]
+
+The user doesn't need to know exactly which package requires which compiler, but is ensured that e.g.  *if* a C++ compiler is needed, then ``llvm`` will be used.
+
+In fact, the conditional dependencies allow the toolchain to be applied to any package regardless of which languages it depends on.
+Similarly, while any spec constraint can be included, it is most useful to use compiler flags, architectures, and conditional dependencies.
+
+Different toolchains could be used independently or even in the same spec.
+If we had a toolchain named ``gcc_all`` that enforces using ``gcc`` for C, C++ and Fortran, we could write:
+
+.. code-block::
+
+   spack install hdf5+fortran%llvm_gfortran ^mpich %gcc_all
+
+to install:
+
+* An ``hdf5`` compiled with ``llvm`` for the C/C++ components, but with the fortran components compiled with ``gfortran``,
+* Built against an MPICH installation compiled entirely with ``gcc`` for C, C++, and Fortran.
+
+.. note::
+
+   Toolchains are currently limited to using only direct dependencies (``%``) in their definition.
+   Transitive dependencies are not allowed.
+
+.. _audit-packages-and-configuration:
+
+===================================
+Auditing Packages and Configuration
+===================================
+
+The ``spack audit`` command:
+
+.. command-output:: spack audit -h
+
+can be used to detect a number of configuration issues. This command detects
+configuration settings that might not be strictly wrong but are not likely
```

Comment:
This is a weird way to introduce the command. You should probably say it detects potential issues with configuration and packages. It's like a linter for config and package.py files.